sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SerializationConstructorsShouldBeSecured.cs

```diff
@@ -0,0 +1,139 @@
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class SerializationConstructorsShouldBeSecured : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4212";
+        private const string MessageFormat = "Secure this serialization constructor.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly AttributeComparer attributeComparer = new AttributeComparer();
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(FindPossibleViolations,
+                SyntaxKind.ConstructorDeclaration);
+        }
+
+        private void FindPossibleViolations(SyntaxNodeAnalysisContext c)
+        {
+            var constructorSyntax = (ConstructorDeclarationSyntax)c.Node;
+            var reportLocation = constructorSyntax?.Identifier.GetLocation();
+            if (reportLocation == null)
+            {
+                return;
+            }
+
+            var serializationConstructor = c.SemanticModel.GetDeclaredSymbol(constructorSyntax);
+            if (!serializationConstructor.IsSerializationConstructor())
+            {
+                return;
+            }
+
+            var classSymbol = serializationConstructor.ContainingType;
+            if (!classSymbol.Implements(KnownType.System_Runtime_Serialization_ISerializable))
+            {
+                return;
+            }
+
+            var isAssemblyIsPartiallyTrusted = c.SemanticModel.Compilation.Assembly
+                    .GetAttributes()
+                    .Any(a => a.AttributeClass.Is(KnownType.System_Security_AllowPartiallyTrustedCallersAttribute));
+            if (!isAssemblyIsPartiallyTrusted)
+            {
+                return;
+            }
+
+            var serializationConstructorAttributes = GetCASAttributes(serializationConstructor).ToHashSet();
+
+            bool isConstructorMissingAttributes = classSymbol.Constructors
+                .SelectMany(m => GetCASAttributes(m))
+                .Any(attr => !serializationConstructorAttributes.Contains(attr, attributeComparer));
+
+            if (isConstructorMissingAttributes)
+            {
+                c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, reportLocation));
+            }
+        }
+
+        private static IEnumerable<AttributeData> GetCASAttributes(IMethodSymbol methodSymbol)
```

Comment:
[ignore me] Ahahah what's the problem with naming the method `GetCodeAccessSecurityAttributes`?

Reply:
The name is almost longer than the implementation (save for inner function) ;)

To me it's more readable. Similar case to `NaN` or `HTML` - why spell it out if the acronyms are well known? And if somebody doesn't know them, some googling is needed anyway. Just writing name in full won't help much.